File: 1824-minimum-sideway-jumps/1824-minimum-sideway-jumps.js
/**
 * @param {number[]} obstacles
 * @return {number}
 */
var minSideJumps = function(obstacles) {
  const dp = [1, 0, 1];
  
  for (let num of obstacles) {
    if (num) dp[num - 1] = Infinity;
    
    for (let i = 0; i < 3; i++) {
      if (i + 1 === num) continue;
      
      dp[i] = Math.min(dp[i], dp[(i + 1) % 3] + 1, dp[(i + 2) % 3] + 1);
    }
  }
  
  return Math.min(...dp);
};

File: 1824-minimum-sideway-jumps/1824-minimum-sideway-jumps.py
class Solution:
    def minSideJumps(self, obstacles: List[int]) -> int:
        dp = [1, 0, 1]
        for num in obstacles:
            if num:
                dp[num - 1] = float('inf')
            for i in range(3):
                if i + 1 != num:
                    dp[i] = min(dp[i], dp[(i + 1) % 3] + 1, dp[(i + 2) % 3] + 1)
        return min(dp)